libs/cache/single_pointer_cache.hpp

```diff
@@ -0,0 +1,72 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_SINGLE_POINTER_CACHE_HPP
+#define IROHA_SINGLE_POINTER_CACHE_HPP
+
+#include <memory>
+#include <mutex>
+
+namespace iroha {
+  namespace cache {
+
+    /**
+     * Thread-safely stores and returns shared pointer to an element of template
+     * type
+     */
+    template <typename DataType>
+    class SinglePointerCache {
+      using DataPointer = std::shared_ptr<DataType>;
+
+      DataPointer stored_data_;
+
+      mutable std::mutex mutex_;
+
+     public:
+      /**
+       * Insert data to the cache
+       * @param pointer to the data to be inserted
+       */
+      void insert(DataPointer data);
+
+      /**
+       * Get data from the cache
+       * @return pointer to the stored data
+       */
+      DataPointer get() const;
+
+      /**
+       * Delete data inside the cache
+       */
+      void release();
+    };
+
+    template <typename DataType>
+    void SinglePointerCache<DataType>::insert(
+        SinglePointerCache::DataPointer data) {
+      std::lock_guard<std::mutex> lock(mutex_);
+
+      stored_data_ = std::move(data);
+    }
+
+    template <typename DataType>
+    typename SinglePointerCache<DataType>::DataPointer
```

Comment:
you can use out of class using to avoid typename

```C++
template <typename Cache>
using CacheData = Cache::DataPointer;
```

Reply:
Should I? It's used only once, so I think it can be left as-is